xapian-core/include/xapian/weight.h

```diff
@@ -590,6 +590,101 @@ class XAPIAN_VISIBILITY_DEFAULT BM25Weight : public Weight {
     double get_maxextra() const;
 };
 
+/// Xapian::Weight subclass implementing the BM25+ probabilistic formula.
+class XAPIAN_VISIBILITY_DEFAULT BM25PlusWeight : public BM25Weight {
+
+  public:
+    /** Construct a BM25PlusWeight.
+     *
+     *  @param delta  A parameter for pseudo tf value to control the scale
+     *		      of the tf lower bound. Delta(δ) can be tuned for e.g
+     *		      from 0.0 to 1.5 but BM25+ can still work effectively
+     *		      across collections with a fixed δ = 1.0. (default 1.0)
+     */
+    BM25PlusWeight(double k1, double k2, double k3, double b, double min_normlen, double delta)
+	: param_k1(k1), param_k2(k2), param_k3(k3), param_b(b),
+	  param_min_normlen(min_normlen), param_delta(delta)
+    {
+	if (param_k1 < 0) param_k1 = 0;
+	if (param_k2 < 0) param_k2 = 0;
+	if (param_k3 < 0) param_k3 = 0;
+	if (param_b < 0) {
+	    param_b = 0;
+	} else if (param_b > 1) {
+	    param_b = 1;
+	}
+	need_stat(COLLECTION_SIZE);
+	need_stat(RSET_SIZE);
+	need_stat(TERMFREQ);
+	need_stat(RELTERMFREQ);
+	need_stat(WDF);
+	need_stat(WDF_MAX);
+	if (param_k2 != 0 || (param_k1 != 0 && param_b != 0)) {
+	    need_stat(DOC_LENGTH_MIN);
+	    need_stat(AVERAGE_LENGTH);
+	}
+	if (param_k1 != 0 && param_b != 0) need_stat(DOC_LENGTH);
+	if (param_k2 != 0) need_stat(QUERY_LENGTH);
+	if (param_k3 != 0) need_stat(WQF);
+	if (param_delta < 0) param_delta = 0;
```

Comment:
The pattern of the existing code is to first clamp all the parameters to valid values, and then do all the unconditional `need_stat()` calls, and then do all the conditional `need_stat()` calls, so this clamping of `param_delta` would fit better higher up.